File: crates/engine/src/tools/write_file.rs
use super::{
    display_path, hash_content, str_arg, FileHashes, Tool, ToolContext, ToolFuture, ToolResult,
};
use serde_json::Value;
use std::collections::HashMap;
use std::path::Path;

pub struct WriteFileTool {
    pub hashes: FileHashes,
}

impl Tool for WriteFileTool {
    fn name(&self) -> &str {
        "write_file"
    }

    fn description(&self) -> &str {
        "Writes a file to the local filesystem. This tool will overwrite the existing file if there is one at the provided path."
    }

    fn parameters(&self) -> Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "file_path": {
                    "type": "string",
                    "description": "The absolute path to the file to write (must be absolute, not relative)"
                },
                "content": {
                    "type": "string",
                    "description": "The content to write to the file"
                }
            },
            "required": ["file_path", "content"]
        })
    }

    fn needs_confirm(&self, args: &HashMap<String, Value>) -> Option<String> {
        Some(display_path(&str_arg(args, "file_path")))
    }

    fn execute<'a>(
        &'a self,
        args: HashMap<String, Value>,
        _ctx: &'a ToolContext<'a>,
    ) -> ToolFuture<'a> {
        Box::pin(async move { tokio::task::block_in_place(|| self.run(&args)) })
    }
}

impl WriteFileTool {
    fn run(&self, args: &HashMap<String, Value>) -> ToolResult {
        let path = str_arg(args, "file_path");
        let content = str_arg(args, "content");

        if Path::new(&path).exists() {
            let has_hash = self.hashes.lock().is_ok_and(|map| map.contains_key(&path));
            if !has_hash {
                return ToolResult {
                    content: "File already exists. Use edit_file to modify existing files, or read_file then write_file to replace.".into(),
                    is_error: true,
                };
            }
        }

        if let Some(parent) = Path::new(&path).parent() {
            if let Err(e) = std::fs::create_dir_all(parent) {
                return ToolResult {
                    content: e.to_string(),
                    is_error: true,
                };
            }
        }

        match std::fs::write(&path, &content) {
            Ok(_) => {
                if let Ok(mut map) = self.hashes.lock() {
                    map.insert(path.clone(), hash_content(&content));
                }
                ToolResult {
                    content: format!("wrote {} bytes to {}", content.len(), display_path(&path)),
                    is_error: false,
                }
            }
            Err(e) => ToolResult {
                content: e.to_string(),
                is_error: true,
            },
        }
    }
}

File: runtime/lua/smelt/tools/write_file.lua
-- Built-in write_file tool. Refuses to overwrite unread files and detects mtime drift.

local UNREAD_OVERWRITE_ERR = "File already exists. Use edit_file to modify existing files, or read_file then write_file to replace."

local function preflight_message(file_path)
  if file_path == "" then
    return nil
  end
  if not smelt.fs.exists(file_path) then
    return nil
  end
  if not smelt.fs.file_state.has(file_path) then
    return UNREAD_OVERWRITE_ERR
  end
  return smelt.fs.file_state.staleness_error(file_path, "file")
end

smelt.tools.register({
  name = "write_file",
  description = "Writes a file to the local filesystem. This tool will overwrite the existing file if there is one at the provided path.",
  override = true,
  permission_defaults = { plan = "deny", apply = "allow" },
  parameters = {
    type = "object",
    properties = {
      file_path = {
        type = "string",
        description = "The absolute path to the file to write (must be absolute, not relative)",
      },
      content = {
        type = "string",
        description = "The content to write to the file",
      },
    },
    required = { "file_path", "content" },
  },
  summary = function(args)
    return smelt.path.display(args.file_path or "")
  end,
  preflight = function(args)
    return preflight_message(args.file_path or "")
  end,
  render = function(args, output, ctx)
    if output.is_error then
      return smelt.layout.text(output.content, { hl_group = "ErrorMsg" })
    end
    return smelt.layout.file_view({
      content = args.content or "",
      path    = args.file_path or "",
    })
  end,
  paths_for_workspace = function(args)
    local p = args.file_path or ""
    return p ~= "" and { p } or {}
  end,
  preview = function(args)
    return smelt.layout.file_view({
      content = args.content or "",
      path    = args.file_path or "",
    })
  end,
  execute = function(args)
    local path = args.file_path or ""
    local content = args.content or ""

    if path == "" then
      return { content = "missing required parameter: file_path", is_error = true }
    end
    if smelt.notebook.is_notebook_path(path) then
      return {
        content = "Cannot use write_file on a Jupyter notebook. Use edit_notebook instead.",
        is_error = true,
      }
    end

    local exists = smelt.fs.exists(path)
    local lock = nil
    if exists then
      if not smelt.fs.file_state.has(path) then
        return { content = UNREAD_OVERWRITE_ERR, is_error = true }
      end
      local stale = smelt.fs.file_state.staleness_error(path, "file")
      if stale then
        return { content = stale, is_error = true }
      end
      local guard, err = smelt.fs.try_flock(path)
      if not guard then
        return { content = err or "could not lock file", is_error = true }
      end
      lock = guard
    end

    local parent = smelt.path.parent(path)
    if parent and parent ~= "" then
      local _, mkdir_err = smelt.fs.mkdir_all(parent)
      if mkdir_err then
        return { content = mkdir_err, is_error = true }
      end
    end

    local len = #content
    local _, write_err = smelt.fs.write(path, content)
    if write_err then
      if lock then lock:release() end
      return { content = write_err, is_error = true }
    end

    smelt.fs.file_state.record_write(path, content)
    if lock then lock:release() end

    return string.format("wrote %d bytes to %s", len, smelt.path.display(path))
  end,
})
